File: decompiled-src/io/legado/app/model/rss/RssParserByRule.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  kotlin.Metadata
 *  kotlin.Pair
 *  kotlin.collections.CollectionsKt
 *  kotlin.jvm.internal.Intrinsics
 *  kotlin.text.StringsKt
 *  org.jetbrains.annotations.NotNull
 *  org.jetbrains.annotations.Nullable
 */
package io.legado.app.model.rss;

import io.legado.app.data.entities.RssArticle;
import io.legado.app.data.entities.RssSource;
import io.legado.app.exception.NoStackTraceException;
import io.legado.app.model.DebugLog;
import io.legado.app.model.analyzeRule.AnalyzeRule;
import io.legado.app.model.analyzeRule.RuleData;
import io.legado.app.model.rss.RssParserDefault;
import io.legado.app.utils.NetworkUtils;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import kotlin.Metadata;
import kotlin.Pair;
import kotlin.collections.CollectionsKt;
import kotlin.jvm.internal.Intrinsics;
import kotlin.text.StringsKt;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Metadata(mv={1, 5, 1}, k=1, xi=48, d1={"\u0000R\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u000b\n\u0000\n\u0002\u0010 \n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0010!\n\u0002\b\u0004\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\b\u00c6\u0002\u0018\u00002\u00020\u0001B\u0007\b\u0002\u00a2\u0006\u0002\u0010\u0002J\u0098\u0001\u0010\u0003\u001a\u0004\u0018\u00010\u00042\u0006\u0010\u0005\u001a\u00020\u00062\u0006\u0010\u0007\u001a\u00020\u00012\u0006\u0010\b\u001a\u00020\t2\b\u0010\n\u001a\u0004\u0018\u00010\u00062\u0006\u0010\u000b\u001a\u00020\f2\u0010\u0010\r\u001a\f\u0012\b\u0012\u00060\u000fR\u00020\t0\u000e2\u0010\u0010\u0010\u001a\f\u0012\b\u0012\u00060\u000fR\u00020\t0\u000e2\u0010\u0010\u0011\u001a\f\u0012\b\u0012\u00060\u000fR\u00020\t0\u000e2\u0010\u0010\u0012\u001a\f\u0012\b\u0012\u00060\u000fR\u00020\t0\u000e2\u0010\u0010\u0013\u001a\f\u0012\b\u0012\u00060\u000fR\u00020\t0\u000e2\b\u0010\u0014\u001a\u0004\u0018\u00010\u0015H\u0002JN\u0010\u0016\u001a\u0016\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u00040\u0018\u0012\u0006\u0012\u0004\u0018\u00010\u00060\u00172\u0006\u0010\u0019\u001a\u00020\u00062\u0006\u0010\u001a\u001a\u00020\u00062\b\u0010\u001b\u001a\u0004\u0018\u00010\u00062\u0006\u0010\u001c\u001a\u00020\u001d2\u0006\u0010\u001e\u001a\u00020\u001f2\b\u0010\u0014\u001a\u0004\u0018\u00010\u0015\u00a8\u0006 "}, d2={"Lio/legado/app/model/rss/RssParserByRule;", "", "()V", "getItem", "Lio/legado/app/data/entities/RssArticle;", "sourceUrl", "", "item", "analyzeRule", "Lio/legado/app/model/analyzeRule/AnalyzeRule;", "variable", "log", "", "ruleTitle", "", "Lio/legado/app/model/analyzeRule/AnalyzeRule$SourceRule;", "rulePubDate", "ruleDescription", "ruleImage", "ruleLink", "debugLog", "Lio/legado/app/model/DebugLog;", "parseXML", "Lkotlin/Pair;", "", "sortName", "sortUrl", "body", "rssSource", "Lio/legado/app/data/entities/RssSource;", "ruleData", "Lio/legado/app/model/analyzeRule/RuleData;", "reader-pro"})
public final class RssParserByRule {
    @NotNull
    public static final RssParserByRule INSTANCE = new RssParserByRule();

    private RssParserByRule() {
    }

    @NotNull
    public final Pair<List<RssArticle>, String> parseXML(@NotNull String sortName, @NotNull String sortUrl, @Nullable String body, @NotNull RssSource rssSource, @NotNull RuleData ruleData, @Nullable DebugLog debugLog) throws Exception {
        boolean bl;
        Object object;
        Intrinsics.checkNotNullParameter((Object)sortName, (String)"sortName");
        Intrinsics.checkNotNullParameter((Object)sortUrl, (String)"sortUrl");
        Intrinsics.checkNotNullParameter((Object)rssSource, (String)"rssSource");
        Intrinsics.checkNotNullParameter((Object)ruleData, (String)"ruleData");
        String sourceUrl = rssSource.getSourceUrl();
        String nextUrl = null;
        CharSequence charSequence = body;
        boolean bl2 = false;
        boolean bl3 = false;
        if (charSequence == null || StringsKt.isBlank((CharSequence)charSequence)) {
            throw new NoStackTraceException(Intrinsics.stringPlus((String)"error_get_web_content: ", (Object)rssSource.getSourceUrl()));
        }
        String ruleArticles = rssSource.getRuleArticles();
        Object object2 = ruleArticles;
        bl3 = false;
        boolean bl4 = false;
        if (object2 == null || StringsKt.isBlank((CharSequence)object2)) {
            object2 = debugLog;
            if (object2 != null) {
                DebugLog.DefaultImpls.log$default((DebugLog)object2, sourceUrl, "\u21d2\u5217\u8868\u89c4\u5219\u4e3a\u7a7a, \u4f7f\u7528\u9ed8\u8ba4\u89c4\u5219\u89e3\u6790", false, 4, null);
            }
            return RssParserDefault.INSTANCE.parseXML(sortName, body, sourceUrl, debugLog);
        }
        bl3 = false;
        List articleList = new ArrayList();
        AnalyzeRule analyzeRule = new AnalyzeRule(ruleData, rssSource, debugLog);
        AnalyzeRule.setContent$default(analyzeRule, body, null, 2, null).setBaseUrl(sortUrl);
        analyzeRule.setRedirectUrl(sortUrl);
        boolean reverse = false;
        if (StringsKt.startsWith$default((String)ruleArticles, (String)"-", (boolean)false, (int)2, null)) {
            reverse = true;
            object = ruleArticles;
            int n = 1;
            bl = false;
            Object object3 = object;
            if (object3 == null) {
                throw new NullPointerException("null cannot be cast to non-null type java.lang.String");
            }
            String string = ((String)object3).substring(n);
            Intrinsics.checkNotNullExpressionValue((Object)string, (String)"(this as java.lang.String).substring(startIndex)");
            ruleArticles = string;
        }
        if ((object = debugLog) != null) {
            DebugLog.DefaultImpls.log$default((DebugLog)object, sourceUrl, "\u250c\u83b7\u53d6\u5217\u8868", false, 4, null);
        }
        List<Object> collections = analyzeRule.getElements(ruleArticles);
        Object object4 = debugLog;
        if (object4 != null) {
            DebugLog.DefaultImpls.log$default((DebugLog)object4, sourceUrl, Intrinsics.stringPlus((String)"\u2514\u5217\u8868\u5927\u5c0f:", (Object)collections.size()), false, 4, null);
        }
        object4 = rssSource.getRuleNextPage();
        bl = false;
        boolean bl5 = false;
        if (!(object4 == null || object4.length() == 0)) {
            object4 = debugLog;
            if (object4 != null) {
                DebugLog.DefaultImpls.log$default((DebugLog)object4, sourceUrl, "\u250c\u83b7\u53d6\u4e0b\u4e00\u9875\u94fe\u63a5", false, 4, null);
            }
            String string = rssSource.getRuleNextPage();
            Intrinsics.checkNotNull((Object)string);
            object4 = string;
            Locale locale = Locale.getDefault();
            Intrinsics.checkNotNullExpressionValue((Object)locale, (String)"getDefault()");
            bl5 = false;
            Object object5 = object4;
            if (object5 == null) {
                throw new NullPointerException("null cannot be cast to non-null type java.lang.String");
            }
            String string2 = ((String)object5).toUpperCase(locale);
            Intrinsics.checkNotNullExpressionValue((Object)string2, (String)"(this as java.lang.String).toUpperCase(locale)");
            if (Intrinsics.areEqual((Object)string2, (Object)"PAGE")) {
                nextUrl = sortUrl;
            } else {
                nextUrl = AnalyzeRule.getString$default(analyzeRule, rssSource.getRuleNextPage(), null, false, 6, null);
                object4 = nextUrl;
                boolean bl6 = false;
                if (object4.length() > 0) {
                    nextUrl = NetworkUtils.INSTANCE.getAbsoluteURL(sortUrl, nextUrl);
                }
            }
            object4 = debugLog;
            if (object4 != null) {
                DebugLog.DefaultImpls.log$default((DebugLog)object4, sourceUrl, Intrinsics.stringPlus((String)"\u2514", (Object)nextUrl), false, 4, null);
            }
        }
        List ruleTitle = AnalyzeRule.splitSourceRule$default(analyzeRule, rssSource.getRuleTitle(), false, 2, null);
        List rulePubDate = AnalyzeRule.splitSourceRule$default(analyzeRule, rssSource.getRulePubDate(), false, 2, null);
        List ruleDescription = AnalyzeRule.splitSourceRule$default(analyzeRule, rssSource.getRuleDescription(), false, 2, null);
        List ruleImage = AnalyzeRule.splitSourceRule$default(analyzeRule, rssSource.getRuleImage(), false, 2, null);
        List ruleLink = AnalyzeRule.splitSourceRule$default(analyzeRule, rssSource.getRuleLink(), false, 2, null);
        String variable = ruleData.getVariable();
        Iterator<Object> iterator = collections.iterator();
        int n = 0;
        while (iterator.hasNext()) {
            int index = n++;
            Object item = iterator.next();
            RssArticle rssArticle = this.getItem(sourceUrl, item, analyzeRule, variable, index == 0, ruleTitle, rulePubDate, ruleDescription, ruleImage, ruleLink, debugLog);
            if (rssArticle == null) continue;
            RssArticle rssArticle2 = rssArticle;
            boolean bl7 = false;
            boolean bl8 = false;
            RssArticle it = rssArticle2;
            boolean bl9 = false;
            it.setSort(sortName);
            it.setOrigin(sourceUrl);
            articleList.add(it);
        }
        if (reverse) {
            CollectionsKt.reverse((List)articleList);
        }
        return new Pair((Object)articleList, (Object)nextUrl);
    }

    private final RssArticle getItem(String sourceUrl, Object item, AnalyzeRule analyzeRule, String variable, boolean log, List<AnalyzeRule.SourceRule> ruleTitle, List<AnalyzeRule.SourceRule> rulePubDate, List<AnalyzeRule.SourceRule> ruleDescription, List<AnalyzeRule.SourceRule> ruleImage, List<AnalyzeRule.SourceRule> ruleLink, DebugLog debugLog) {
        RssArticle rssArticle = new RssArticle(null, null, null, 0L, null, null, null, null, null, false, variable, 1023, null);
        analyzeRule.setRuleData(rssArticle);
        AnalyzeRule.setContent$default(analyzeRule, item, null, 2, null);
        Object object = debugLog;
        if (object != null) {
            object.log(sourceUrl, "\u250c\u83b7\u53d6\u6807\u9898", log);
        }
        rssArticle.setTitle(AnalyzeRule.getString$default(analyzeRule, ruleTitle, null, false, 6, null));
        object = debugLog;
        if (object != null) {
            object.log(sourceUrl, Intrinsics.stringPlus((String)"\u2514", (Object)rssArticle.getTitle()), log);
        }
        object = debugLog;
        if (object != null) {
            object.log(sourceUrl, "\u250c\u83b7\u53d6\u65f6\u95f4", log);
        }
        rssArticle.setPubDate(AnalyzeRule.getString$default(analyzeRule, rulePubDate, null, false, 6, null));
        object = debugLog;
        if (object != null) {
            object.log(sourceUrl, Intrinsics.stringPlus((String)"\u2514", (Object)rssArticle.getPubDate()), log);
        }
        object = debugLog;
        if (object != null) {
            object.log(sourceUrl, "\u250c\u83b7\u53d6\u63cf\u8ff0", log);
        }
        object = ruleDescription;
        boolean bl = false;
        boolean bl2 = false;
        if (object == null || object.isEmpty()) {
            rssArticle.setDescription(null);
            object = debugLog;
            if (object != null) {
                object.log(sourceUrl, "\u2514\u63cf\u8ff0\u89c4\u5219\u4e3a\u7a7a\uff0c\u5c06\u4f1a\u89e3\u6790\u5185\u5bb9\u9875", log);
            }
        } else {
            rssArticle.setDescription(AnalyzeRule.getString$default(analyzeRule, ruleDescription, null, false, 6, null));
            object = debugLog;
            if (object != null) {
                object.log(sourceUrl, Intrinsics.stringPlus((String)"\u2514", (Object)rssArticle.getDescription()), log);
            }
        }
        object = debugLog;
        if (object != null) {
            object.log(sourceUrl, "\u250c\u83b7\u53d6\u56fe\u7247url", log);
        }
        rssArticle.setImage(AnalyzeRule.getString$default(analyzeRule, ruleImage, null, true, 2, null));
        object = debugLog;
        if (object != null) {
            object.log(sourceUrl, Intrinsics.stringPlus((String)"\u2514", (Object)rssArticle.getImage()), log);
        }
        object = debugLog;
        if (object != null) {
            object.log(sourceUrl, "\u250c\u83b7\u53d6\u6587\u7ae0\u94fe\u63a5", log);
        }
        rssArticle.setLink(NetworkUtils.INSTANCE.getAbsoluteURL(sourceUrl, AnalyzeRule.getString$default(analyzeRule, ruleLink, null, false, 6, null)));
        object = debugLog;
        if (object != null) {
            object.log(sourceUrl, Intrinsics.stringPlus((String)"\u2514", (Object)rssArticle.getLink()), log);
        }
        if (StringsKt.isBlank((CharSequence)rssArticle.getTitle())) {
            return null;
        }
        return rssArticle;
    }
}

File: best-of-3/src/io/legado/app/model/rss/RssParserByRule.kt
package io.legado.app.model.rss

import io.legado.app.data.entities.RssArticle
import io.legado.app.data.entities.RssSource
import io.legado.app.exception.NoStackTraceException
import io.legado.app.model.DebugLog
import io.legado.app.model.analyzeRule.AnalyzeRule
import io.legado.app.model.analyzeRule.RuleData
import io.legado.app.model.analyzeRule.AnalyzeRule.SourceRule
import io.legado.app.utils.NetworkUtils
import java.util.ArrayList
import java.util.Locale
import kotlin.jvm.internal.Intrinsics

public object RssParserByRule {
   @Throws(java/lang/Exception::class)
   public fun parseXML(sortName: String, sortUrl: String, body: String?, rssSource: RssSource, ruleData: RuleData, debugLog: DebugLog?): Pair<
         MutableList<RssArticle>,
         String?
      > {
      val sourceUrl: java.lang.String = rssSource.getSourceUrl();
      var nextUrl: java.lang.String = null;
      if (body == null || StringsKt.isBlank(body)) {
         throw new NoStackTraceException(Intrinsics.stringPlus("error_get_web_content: ", rssSource.getSourceUrl()));
      } else {
         var var30: java.lang.String = rssSource.getRuleArticles();
         if (var30 == null || StringsKt.isBlank(var30)) {
            if (debugLog != null) {
               DebugLog.DefaultImpls.log$default(debugLog, sourceUrl, "⇒列表规则为空, 使用默认规则解析", false, 4, null);
            }

            return RssParserDefault.INSTANCE.parseXML(sortName, body, sourceUrl, debugLog);
         } else {
            val var32: java.util.List = new ArrayList();
            val var35: AnalyzeRule = new AnalyzeRule(ruleData, rssSource, debugLog);
            AnalyzeRule.setContent$default(var35, body, null, 2, null).setBaseUrl(sortUrl);
            var35.setRedirectUrl(sortUrl);
            var var36: Boolean = false;
            if (StringsKt.startsWith$default(var30, "-", false, 2, null)) {
               var36 = true;
               if (var30 == null) {
                  throw new NullPointerException("null cannot be cast to non-null type java.lang.String");
               }

               val var10000: java.lang.String = var30.substring(1);
               var30 = var10000;
            }

            if (debugLog != null) {
               DebugLog.DefaultImpls.log$default(debugLog, sourceUrl, "┌获取列表", false, 4, null);
            }

            val collections: java.util.List = var35.getElements(var30);
            if (debugLog != null) {
               DebugLog.DefaultImpls.log$default(debugLog, sourceUrl, Intrinsics.stringPlus("└列表大小:", collections.size()), false, 4, null);
            }

            val var37: java.lang.CharSequence = rssSource.getRuleNextPage();
            if (var37 != null && var37.length() != 0) {
               if (debugLog != null) {
                  DebugLog.DefaultImpls.log$default(debugLog, sourceUrl, "┌获取下一页链接", false, 4, null);
               }

               var var47: java.lang.String = rssSource.getRuleNextPage();
               val var42: Locale = Locale.getDefault();
               if (var47 == null) {
                  throw new NullPointerException("null cannot be cast to non-null type java.lang.String");
               }

               var47 = var47.toUpperCase(var42);
               if (var47 == "PAGE") {
                  nextUrl = sortUrl;
               } else {
                  nextUrl = AnalyzeRule.getString$default(var35, rssSource.getRuleNextPage(), null, false, 6, null);
                  if (nextUrl.length() > 0) {
                     nextUrl = NetworkUtils.INSTANCE.getAbsoluteURL(sortUrl, nextUrl);
                  }
               }

               if (debugLog != null) {
                  DebugLog.DefaultImpls.log$default(debugLog, sourceUrl, Intrinsics.stringPlus("└", nextUrl), false, 4, null);
               }
            }

            val var40: java.util.List = AnalyzeRule.splitSourceRule$default(var35, rssSource.getRuleTitle(), false, 2, null);
            val var44: java.util.List = AnalyzeRule.splitSourceRule$default(var35, rssSource.getRulePubDate(), false, 2, null);
            val var46: java.util.List = AnalyzeRule.splitSourceRule$default(var35, rssSource.getRuleDescription(), false, 2, null);
            val ruleImage: java.util.List = AnalyzeRule.splitSourceRule$default(var35, rssSource.getRuleImage(), false, 2, null);
            val ruleLink: java.util.List = AnalyzeRule.splitSourceRule$default(var35, rssSource.getRuleLink(), false, 2, null);
            val variable: java.lang.String = ruleData.getVariable();
            val var20: java.util.Iterator = collections.iterator();
            var var21: Int = 0;

            while (var20.hasNext()) {
               val var24: RssArticle = this.getItem(sourceUrl, var20.next(), var35, variable, var21++ == 0, var40, var44, var46, ruleImage, ruleLink, debugLog);
               if (var24 != null) {
                  var24.setSort(sortName);
                  var24.setOrigin(sourceUrl);
                  var32.add(var24);
               }
            }

            if (var36) {
               CollectionsKt.reverse(var32);
            }

            return new Pair<>(var32, nextUrl);
         }
      }
   }

   private fun getItem(
      sourceUrl: String,
      item: Any,
      analyzeRule: AnalyzeRule,
      variable: String?,
      log: Boolean,
      ruleTitle: List<SourceRule>,
      rulePubDate: List<SourceRule>,
      ruleDescription: List<SourceRule>,
      ruleImage: List<SourceRule>,
      ruleLink: List<SourceRule>,
      debugLog: DebugLog?
   ): RssArticle? {
      val rssArticle: RssArticle = new RssArticle(null, null, null, 0L, null, null, null, null, null, false, variable, 1023, null);
      analyzeRule.setRuleData(rssArticle);
      AnalyzeRule.setContent$default(analyzeRule, item, null, 2, null);
      if (debugLog != null) {
         debugLog.log(sourceUrl, "┌获取标题", log);
      }

      rssArticle.setTitle(AnalyzeRule.getString$default(analyzeRule, ruleTitle, null, false, 6, null));
      if (debugLog != null) {
         debugLog.log(sourceUrl, Intrinsics.stringPlus("└", rssArticle.getTitle()), log);
      }

      if (debugLog != null) {
         debugLog.log(sourceUrl, "┌获取时间", log);
      }

      rssArticle.setPubDate(AnalyzeRule.getString$default(analyzeRule, rulePubDate, null, false, 6, null));
      if (debugLog != null) {
         debugLog.log(sourceUrl, Intrinsics.stringPlus("└", rssArticle.getPubDate()), log);
      }

      if (debugLog != null) {
         debugLog.log(sourceUrl, "┌获取描述", log);
      }

      if (ruleDescription == null || ruleDescription.isEmpty()) {
         rssArticle.setDescription(null);
         if (debugLog != null) {
            debugLog.log(sourceUrl, "└描述规则为空，将会解析内容页", log);
         }
      } else {
         rssArticle.setDescription(AnalyzeRule.getString$default(analyzeRule, ruleDescription, null, false, 6, null));
         if (debugLog != null) {
            debugLog.log(sourceUrl, Intrinsics.stringPlus("└", rssArticle.getDescription()), log);
         }
      }

      if (debugLog != null) {
         debugLog.log(sourceUrl, "┌获取图片url", log);
      }

      rssArticle.setImage(AnalyzeRule.getString$default(analyzeRule, ruleImage, null, true, 2, null));
      if (debugLog != null) {
         debugLog.log(sourceUrl, Intrinsics.stringPlus("└", rssArticle.getImage()), log);
      }

      if (debugLog != null) {
         debugLog.log(sourceUrl, "┌获取文章链接", log);
      }

      rssArticle.setLink(NetworkUtils.INSTANCE.getAbsoluteURL(sourceUrl, AnalyzeRule.getString$default(analyzeRule, ruleLink, null, false, 6, null)));
      if (debugLog != null) {
         debugLog.log(sourceUrl, Intrinsics.stringPlus("└", rssArticle.getLink()), log);
      }

      return if (StringsKt.isBlank(rssArticle.getTitle())) null else rssArticle;
   }
}
